docs/debug.md

```diff
@@ -0,0 +1,173 @@
+
+**For pre requirements, see the [Prerequisites](/docs/contributing.md#prerequisites) section on the Azure Key Vault CSI Driver Provider for Azure.**
+
+### Table of Contents
+- [Azure Key Vault Creation](#azure-key-vault-creation)
+- [Service Principal Generation](#service-principal-generation)
+- [Set Up Debug Directory](#set-up-debug-directory)
+  - [launch.json](#launch.json)
+  - [parameters.yaml](#parameters.yaml)
+  - [build-args.sh](#build-args)
+- [Running the Debugger and Testing](#running-the-debugger-and-testing)
+  - [Break Points](#breakpoints)
+- [Testing with Created Azure Service Principal](#testing-with-created-azure-service-principal)
+
+## Azure Key Vault Creation
+
+The Azure Key Vault will be used to store secrets, keys, credentials that we provide.
+
+1. Create a Resource Group that will contain the Azure Key Vault
+    ```bash
+      RESOURCEGROUP=AzureProviderDemoGroup
+      az group create -n $RESOURCEGROUP --location westus
+    ```
+2. Create the Azure Key Vault
+    - location will be set to westus. This Resource will be placed inside our previously created Resource Group `AzureProviderDemoGroup`.
+    ```bash
+    # $RANDOM will generate a random between 0 and 32767
+    KEYVAULT_NAME=keyvault-demo-$(echo $RANDOM)
+
+    az keyvault create -n $KEYVAULT_NAME -g $RESOURCEGROUP --location westus
+    ```
+
+## Service Principal Generation
+
+A service principal needs to be generated to create a user with limited permissions to access the keys in the keyvault. We will set keys in the Azure Key Vault through our **original** admin privileges, and then **Read** the keys through the **generated service principal permissions.**
+
+1. Receive the Key Vault Resource ID
+    ```bash
+    KEYVAULT_RESOURCE_ID=$(az keyvault show -n $KEYVAULT_NAME --query id -o tsv)
+    ```
+
+2. Generate The Service Principal and query the Client ID
+    ```bash
+    CLIENT_ID=$(az ad sp create-for-rbac  --name $KEYVAULT_NAME --role Reader --scopes $KEYVAULT_RESOURCE_ID --query appId -o tsv)
+    ```
+    - this is your KEYVAULT_CLIENT_ID
+3. Reset the Service Principal Password(Client Secret) and store it in `CLIENT_SECRET`
+    ```bash
+    CLIENT_SECRET=$(az ad sp credential reset --name $CLIENT_ID --credential-description "APClientSecret" --query password -o tsv)
+    ```
+    - This is your  `KEYVAULT_CLIENT_SECRET`
+4. Set the Azure Key Vault Policy to `Get` (Read Only) using the Service Principal just now created
+
+        az keyvault set-policy -n $KEYVAULT_NAME --secret-permissions get --spn $CLIENT_ID
+
+5. While still under admin privileges, add the `CLIENT_SECRET` and `CLIENT_ID` to the Key Vault as secrets.
+      ```bash
+      az keyvault secret set --vault-name $KEYVAULT_NAME --name "clientSecret" --value $CLIENT_SECRET
+      az keyvault secret set --vault-name $KEYVAULT_NAME --name "clientId" --value $CLIENT_ID
+      ```
+
+## Set Up Debug Directory
+
```

Comment:
List all the artifacts in the debug folder and provide a short paragraph on what their functions are. Then dive into the details and how to use them in the following sections below.